alphastats/llm/llm_functions.py

```diff
@@ -1,15 +1,119 @@
 """Module for defining assistant functions for the ChatGPT model."""
 
-from typing import Dict, List
+from typing import Dict, List, Tuple, Union
 
 import pandas as pd
+import streamlit as st
 
 from alphastats.dataset.dataset import DataSet
+from alphastats.gui.utils.ui_helper import StateKeys
 from alphastats.llm.enrichment_analysis import get_enrichment_data
-from alphastats.llm.uniprot_utils import get_gene_function
+from alphastats.llm.uniprot_utils import (
+    format_uniprot_annotation,
+    get_annotations_for_feature,
+)
+
+
+def get_annotation_from_store_by_feature_list(
+    features: list, annotation_store: dict
+) -> Union[dict, str, None]:
+    """
+    Retrieve the annotation from the store based on a list of feauture ids.
+    The first matching feature found in the annotation store, is returned. Don't use this for disparate features, it is meant to be used for features that have the same representation.
+    Args:
+        features (list): A list of features to search for.
+        annotation_store (dict): A dictionary storing annotations for features.
+    Returns:
+        Union[dict, str, None]: The annotation corresponding to the feature representation if found, otherwise None.
+    """
+    for feature in features:
+        if feature in annotation_store:
+            return annotation_store[feature]
+    return None
+
+
+def get_annotation_from_uniprot_by_feature_list(
+    features: list,
+) -> Tuple[Union[dict, str], str]:
+    """
+    Retrieve annotation from UniProt based on a list of features.
+    It retrieves the annotation for the most appropriate feature from UniProt. Most appropriate is defined as 1. the only feature, 2. any feature if they only differ in isoform ids, 3. the feature with the most base ids, as this is most likely to contain a well annotated protein.
+    Args:
+        features (list): A list of features to search for.
+    Returns:
+        Union[dict, str]: The annotation for the identified feature. The return type can be a dictionaryn or a string depending on the annotation retrieved.
+        feature (str): The feature that was used to retrieve the annotation.
+    """
+    if len(features) == 0:
+        raise ValueError("No features provided.")
+    elif len(features) == 1:
+        feature = features[0]
+    else:
+        baseid_sets = []
+        for feature in features:
+            baseid_sets.append(
+                set([identifier.split("-")[0] for identifier in feature.split(";")])
+            )
+        if len(set.union(*baseid_sets)) == len(set.intersection(*baseid_sets)):
+            feature = features[0]
+        else:
+            feature = features[baseid_sets.index(max(baseid_sets, key=len))]
+    annotation = get_annotations_for_feature(feature)
```

Comment:
maybe we could always return a dict, even when there is a failure?

Reply:
I actually prefer, that along the way strings are returned and passed on, as this might inform the LLM as to why there was no information for a specific entry, rather than just providing nothing. We could then include in the system prompt, that certain annotations should lead to specific outcomes, like rating a protein more lowly in it's weight on the biological interpretation.